complex/ft_domain.c

```diff
@@ -191,14 +192,53 @@ static int ft_setup_xcontrol_bufs(struct ft_xcontrol *ctrl)
 		memset(ctrl->buf, 0, size);
 	}
 
-	if ((fabric_info->mode & FI_LOCAL_MR) && !ctrl->mr) {
-		ret = fi_mr_reg(domain, ctrl->buf, size, FI_RECV | FI_SEND,
+	if (!ctrl->res_buf) {
+		ctrl->res_buf = calloc(1, size);
+		if (!ctrl->res_buf)
+			return -FI_ENOMEM;
+	} else {
+		memset(ctrl->res_buf, 0, size);
+	}
+
+	if (!ctrl->comp_buf) {
+		ctrl->comp_buf = calloc(1, size);
+		if (!ctrl->comp_buf)
+			return -FI_ENOMEM;
+	} else {
+		memset(ctrl->comp_buf, 0, size);
+	}
+
+	if (fabric_info->mode & FI_LOCAL_MR) {
+		if (!ctrl->mr) {
+			ret = fi_mr_reg(domain, ctrl->buf, size, FI_RECV | FI_SEND,
 				0, 0, 0, &ctrl->mr, NULL);
-		if (ret) {
-			FT_PRINTERR("fi_mr_reg", ret);
-			return ret;
+			if (ret) {
+				FT_PRINTERR("fi_mr_reg", ret);
+				return ret;
+			}
+			ctrl->memdesc = fi_mr_desc(ctrl->mr);
+		}
+
+		access = FI_READ | FI_WRITE | FI_REMOTE_READ | FI_REMOTE_WRITE;
+		if (!ctrl->res_mr) {
+			ret = fi_mr_reg(domain, ctrl->res_buf, size, access,
+				0, 0, 0, &ctrl->res_mr, NULL);
+			if (ret) {
+				FT_PRINTERR("fi_mr_reg", ret);
+				return ret;
+			}
+			ctrl->res_memdesc = fi_mr_desc(ctrl->res_mr);
+		}
+
+		if (!ctrl->comp_mr) {
+			ret = fi_mr_reg(domain, ctrl->comp_buf, size, access,
+				0, 0, 0, &ctrl->comp_mr, NULL);
+			if (ret) {
+				FT_PRINTERR("fi_mr_reg", ret);
+				return ret;
+			}
+			ctrl->comp_memdesc = fi_mr_desc(ctrl->mr);
```

Comment:
This function is getting too long.  Please create a new function that initializes the atomic portions